### PR TITLE
Gcds pagination

### PR DESCRIPTION
**Pull request by ethanWallace**
# Summary | Résumé

Add the `gcds-pagination` component.

``` html
<gcds-pagination
  display="list"                // simple *required
  label=""                      // string *required
 
  // type=”simple” specific attributes
  previous-href=""              // string
  previous-label=""             // string
  next-href=""                  // string
  next-label=""                 // string
 
  // type=”list” specific attributes
  total-pages=""                // number
  current-page=""               // number
  url={}                        // object | string
 
  pageChangeHandler={}          // function
>
</gcds-pagination>
```

### Simple

With predefined pages use the `simple` display. Set the href and label using `previous-href`/`next-href` and `previous-label`/`next-label`.

### List

For a pagination list, use the `list` display. Setting the `total-pages` and the `current-page` will render a responsive pagination list.

### CODE REVIEW

File: packages/web/src/components/gcds-pagination/gcds-pagination.css [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    padding: 0;
+    width: fit-content;
+    list-style: none;
+
+    li {
+      margin: var(--gcds-spacing-100);
+
+      a {
+        color: var(--var-pagination-default-text);
+        font-weight: 800;
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.css:14]: Can you use one of the existing font-weight tokens here? I don't think we are using `font-weight: 800;` anywhere else so maybe use 700 aka `--gcds-font-weights-bold` instead?

File: packages/web/src/components/gcds-pagination/gcds-pagination.css [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  
+            box-shadow: 0 0 0 .125rem var(--gcds-pagination-focus-shadow);
+            outline: .1875rem solid var(--gcds-pagination-focus-outline);
+            outline-offset: .125rem;
+          }
+        }
+      }
+    }
+
+    @media screen and (max-width: 22.5rem) {
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.css:48]: Can we try to use `em` for media queries to be consistent?

File: packages/web/src/components/gcds-pagination/gcds-pagination.css [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        & > span.gcds-pagination-list-ellipses {
+          height: 2.75rem;
+          min-width: 2.75rem;
+          display: flex;
+          justify-content: center;
+          align-items: center;
+
+          &.gcds-pagination-end-button,
+          &.gcds-pagination-end-button-mobile {
+            padding: 0.7rem 0.5rem;
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.css:72]: Could we use `--gcds-spacing-300` and `--gcds-spacing-200` here instead?

File: packages/web/src/components/gcds-pagination/gcds-pagination.css [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+            width: auto;
+            min-width: auto;
+            height:auto;
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.css:75]: Missed a space after the colon.

File: packages/web/src/components/gcds-pagination/gcds-pagination.css [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        &.gcds-pagination-mobile-prevnext {
+          display: none;
+        }
+
+        &.gcds-pagination-list-mobile-ellipses,
+        &.gcds-pagination-list-mobile-ellipses-,
+        &.gcds-pagination-list-mobile-ellipses-360,
+        &.gcds-pagination-list-mobile-ellipses-420,
+        &.gcds-pagination-list-mobile-ellipses-460,
+        &.gcds-pagination-list-mobile-ellipses-488 {
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.css:97]: I think you can summarize all of these classes with:
`&[class^='gcds-pagination-list-mobile-ellipses']`

File: packages/web/src/components/gcds-pagination/gcds-pagination.css [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+          display: none;
+        }
+
+        /* Mobile styling */
+        @media screen and (max-width: 48rem) {
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.css:102]: Same as above, let's try to use `em` for media queries.

File: packages/web/src/components/gcds-pagination/gcds-pagination.css [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+      }
+    }
+
+    &.gcds-pagination-simple {
+      li {
+        margin-bottom: var(--gcds-spacing-200);
+
+        a {
+          display: grid;
+          padding: 0.7rem 0.5rem;
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.css:174]: Could we use --gcds-spacing-300 and --gcds-spacing-200 here instead?

File: packages/web/src/components/gcds-pagination/gcds-pagination.tsx [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+            class="gcds-pagination-simple"
+          >
+            <li>
+              <a
+                href={previousHref}
+                aria-label={`${I18N[lang].previousPage}${previousLabel ? `: ${previousLabel}` : ""}`}
+                onClick={(e) => this.onPageChange(e)}
+              >
+                <gcds-icon class="gcds-pagination-arrow-left" name="arrow-left"></gcds-icon>
+                <b>
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.tsx:307]: Can you switch this to a `<strong>` tag? The `<b>` tag is kinda outdated and shouldn't really be used anymore.

File: packages/web/src/components/gcds-pagination/gcds-pagination.tsx [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+              </a>
+            </li>
+            <li>
+              <a
+                href={nextHref}
+                aria-label={`${I18N[lang].nextPage}${nextLabel ? `: ${nextLabel}` : ""}`}
+                onClick={(e) => this.onPageChange(e)}
+              >
+                <gcds-icon class="gcds-pagination-arrow-right" name="arrow-right"></gcds-icon>
+                <b>
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.tsx:322]: Same comment about the `<b>` tag.

File: packages/web/src/components/gcds-pagination/gcds-pagination.css [hunk 10]
```diff
@@ -0,0 +1,199 @@
+:host {
+  display: block;
+
+  ul {
+    padding: 0;
+    width: fit-content;
+    list-style: none;
+
+    li {
+      margin: var(--gcds-spacing-100);
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.css:10]: Remove this margin for the previous and next buttons when they are not displayed. It will give the component a little bit more space on smaller screen sizes.

File: packages/web/src/components/gcds-pagination/gcds-pagination.css [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+            width: auto;
+            min-width: auto;
+            height:auto;
+  
+            gcds-icon {
+              &.gcds-pagination-arrow-left {
+                margin-right: var(--gcds-spacing-200);
+              }
+              &.gcds-pagination-arrow-right {
+                margin-left: var(--gcds-spacing-200);
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.css:82]: The `gcds-icon` has a prop for adding left or right margin that you can use instead of having to add extra CSS.

File: packages/web/src/components/gcds-pagination/utils/render.tsx [hunk 12]
```diff
@@ -0,0 +1,227 @@
+/**
+ * Function to constuct href attribute from url object
+ * Also looks for ::offset and ::match to increment query string values
+ */
+function constructHref(el, page: number, end?: "next" | "previous" | null) {
+  var fragment = "";
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/utils/render.tsx:6]: Can we use `const` / `let` instead of var everwhere?

File: packages/web/src/components/gcds-pagination/utils/render.tsx [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+
+  return href;
+}
+
+/**
+ * Function to constuct classes to help with mobile sizing
+ */
+function constructClasses(page: number, current: number, total: number) {
+  if (total <= 6) {
+
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/utils/render.tsx:83]: Is there code missing here or can we remove this part?

File: packages/web/src/components/gcds-pagination/utils/render.tsx [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+  var href = `${qs}${fragment}`;
+
+  return href;
+}
+
+/**
+ * Function to constuct classes to help with mobile sizing
+ */
+function constructClasses(page: number, current: number, total: number) {
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/utils/render.tsx:81]: I think we can handle this with CSS and remove some of the complicated logic.

File: packages/web/src/components/gcds-pagination/gcds-pagination.tsx [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+    var nextEllipses = false;
+
+    for (var i = 1; i <= this.totalPages; i++) {
+
+      // Current page
+      if (i == this.currentPage) {
+        this.listitems.push(this.configurePaginationStep(i));
+      }
+      // Total pages less than 10
+      else if (this.totalPages < 10) {
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.tsx:154]: I think we can remove all these switch cases and handle the breakpoints with CSS instead.

**Review by melaniebmn — CHANGES_REQUESTED** [review]
Left some notes and things to work out. Please reach out with any questions!

**Comment by ethanWallace** [comment]
@melaniebmn I wasn't able to fully implement what we talked about but I was able to trim down some of the JS and move a little more of the responsive logic into CSS.

File: packages/web/src/components/gcds-pagination/gcds-pagination.css [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+          &.gcds-pagination-mobile-prevnext {
+            display: block;
+          }
+        }
+
+        /* Mobile breakpoints for responsive rendering */
+        @media screen and (max-width: 33.75em) {
+          &.gcds-pagination-list-breakpoint-488 {
+            display: none;
+          }
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.css:114]: This is very nitpicky, but for the sake of consistency, can you please add a line break after each bracket? It makes it a little bit easier to read. Thank you :)

File: packages/web/src/components/gcds-pagination/gcds-pagination.css [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+          &.gcds-pagination-list-mobile-ellipses:has(+ .gcds-pagination-list-breakpoint-488),
+          &.gcds-pagination-list-breakpoint-488 + .gcds-pagination-list-mobile-ellipses {
+            display: list-item;
+          }
+        }
+        @media screen and (max-width: 31em) {
+          &.gcds-pagination-list-breakpoint-460 {
+            display: none;
+            margin:0;
+          }
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.css:124]: Same comment, and please also add a space after the colon.

File: packages/web/src/components/gcds-pagination/gcds-pagination.tsx [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+        :
+          `${I18N[this.lang].previousPage}: ${I18N[this.lang].pageNumberOf.replace('{#}', --page).replace('{total}', this.totalPages).replace('{label}', this.label)}`,
+    }
+
+    if (page == this.currentPage && !end) {
+      linkAttrs['aria-current'] = "page";
+    }
+
+    if (end == "next") {
+      return(<li><a  {...linkAttrs} class={!mobile ? "gcds-pagination-end-button" : "gcds-pagination-end-button-mobile"}>{I18N[this.lang].next}<gcds-icon margin-left="spacing-200" name="arrow-right"></gcds-icon></a></li>);
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.tsx:120]: Can you format this a bit more so that it's easier to read? Something like:
`
return(
        <li>
          <a
            {...linkAttrs}
            class={!mobile ? "gcds-pagination-end-button" : "gcds-pagination-end-button-mobile"}
          >
            {I18N[this.lang].next}
            <gcds-icon margin-left="spacing-200" name="arrow-right"></gcds-icon>
          </a>
        </li>
      );
`

File: packages/web/src/components/gcds-pagination/gcds-pagination.tsx [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+    }
+    else if (end == "previous") {
+      return(<li><a  {...linkAttrs} class={!mobile ? "gcds-pagination-end-button" : "gcds-pagination-end-button-mobile"}><gcds-icon margin-right="spacing-200" name="arrow-left"></gcds-icon>{I18N[this.lang].previous}</a></li>);
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.tsx:123]: Same comment about formatting.

File: packages/web/src/components/gcds-pagination/gcds-pagination.tsx [hunk 20]
```diff
@@ -0,0 +1,265 @@
...
+    }
+    else {
+      return (<li class={page != 1 && page != this.totalPages ?  constructClasses(page, this.currentPage, this.totalPages) : "" }><a {...linkAttrs}>{page}</a></li>);
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.tsx:126]: Same comment about formatting.

File: packages/web/src/components/gcds-pagination/gcds-pagination.tsx (hunk 19 above)
Comment (melaniebmn) [reply packages/web/src/components/gcds-pagination/gcds-pagination.tsx:123]: Actually, looking at both returns, you could use the same `<li>` and `<a>` code for both and then use an if statement to load in the correct content.

File: packages/web/src/components/gcds-pagination/gcds-pagination.tsx [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+    }
+
+    // Flags to see if ellipses already rendered
+    let previousEllipses = false;
+    let nextEllipses = false;
+
+    for (let i = 1; i <= this.totalPages; i++) {
+
+      if (i == 2 && this.currentPage < 6 && this.currentPage > 3 && this.totalPages > 9) {
+        this.listitems.push(<li class={`gcds-pagination-list-mobile-ellipses`} aria-hidden="true"><span class="gcds-pagination-list-ellipses">...</span></li>);
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.tsx:150]: Same comment about formatting.

File: packages/web/src/components/gcds-pagination/gcds-pagination.tsx [hunk 22]
```diff
@@ -0,0 +1,265 @@
...
+      } else if (i == 2 && this.totalPages < 10 && this.totalPages > 6 && this.currentPage > 3) {
+        this.listitems.push(<li class={`gcds-pagination-list-mobile-ellipses`} aria-hidden="true"><span class="gcds-pagination-list-ellipses">...</span></li>);
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.tsx:152]: Same comment about formatting.

File: packages/web/src/components/gcds-pagination/gcds-pagination.tsx [hunk 23]
```diff
@@ -0,0 +1,265 @@
...
+      } else if (!previousEllipses && i < (this.currentPage-2)) {
+        this.listitems.push(<li aria-hidden="true"><span class="gcds-pagination-list-ellipses">...</span></li>);
+        previousEllipses = true;
+      } else if (!nextEllipses && i > (this.currentPage+2) && i < this.totalPages) {
+        this.listitems.push(<li aria-hidden="true"><span class="gcds-pagination-list-ellipses">...</span></li>);
+        nextEllipses = true;
+      }
+
+      if (i == this.totalPages - 1 && this.currentPage > this.totalPages - 5 && this.currentPage < this.totalPages - 2 && this.totalPages > 9) {
+        this.listitems.push(<li class={`gcds-pagination-list-mobile-ellipses`} aria-hidden="true"><span class="gcds-pagination-list-ellipses">...</span></li>);
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.tsx:176]: Same comment about formatting.

File: packages/web/src/components/gcds-pagination/gcds-pagination.tsx [hunk 24]
```diff
@@ -0,0 +1,265 @@
...
+      } else if (i >= (this.currentPage-2) && i <= (this.currentPage+2)) {
+        this.listitems.push(this.configurePaginationStep(i));
+      } else if ((this.currentPage <= 5 && i <= 7) || (this.currentPage >= this.totalPages - 4 && i >= this.totalPages - 6)) {
+        this.listitems.push(this.configurePaginationStep(i));
+      } else if ((this.currentPage == 5 && i == 2) || (this.currentPage == this.totalPages -4 && i == this.totalPages - 1)) {
+        this.listitems.push(this.configurePaginationStep(i));
+      } else if (this.totalPages < 10) {
+        this.listitems.push(this.configurePaginationStep(i));
+      } else if (!previousEllipses && i < (this.currentPage-2)) {
+        this.listitems.push(<li aria-hidden="true"><span class="gcds-pagination-list-ellipses">...</span></li>);
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.tsx:168]: Same comment about formatting.

File: packages/web/src/components/gcds-pagination/gcds-pagination.tsx [hunk 25]
```diff
@@ -0,0 +1,265 @@
...
+        previousEllipses = true;
+      } else if (!nextEllipses && i > (this.currentPage+2) && i < this.totalPages) {
+        this.listitems.push(<li aria-hidden="true"><span class="gcds-pagination-list-ellipses">...</span></li>);
+        nextEllipses = true;
+      }
+
+      if (i == this.totalPages - 1 && this.currentPage > this.totalPages - 5 && this.currentPage < this.totalPages - 2 && this.totalPages > 9) {
+        this.listitems.push(<li class={`gcds-pagination-list-mobile-ellipses`} aria-hidden="true"><span class="gcds-pagination-list-ellipses">...</span></li>);
+      } else if (i == this.totalPages - 1 && this.totalPages < 10 && this.totalPages > 6 && this.currentPage < this.totalPages - 2) {
+        this.listitems.push(<li class={`gcds-pagination-list-mobile-ellipses`} aria-hidden="true"><span class="gcds-pagination-list-ellipses">...</span></li>);
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.tsx:178]: Same comment about formatting.

File: packages/web/src/components/gcds-pagination/utils/render.tsx [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+ * Function to constuct href attribute from url object
+ * Also looks for ::offset and ::match to increment query string values
+ */
+function constructHref(el, page: number, end?: "next" | "previous" | null) {
+  let fragment = "";
+  let qs = "";
+
+  if (el.url) {
+    let count = 0;
+    for (const key in el.url.queryStrings) {
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/utils/render.tsx:11]: Sorry very nit-picky again, but for consistency: please add a space before and after each if statement, for loops, etc.

File: packages/web/src/components/gcds-pagination/gcds-pagination.tsx [hunk 27]
```diff
@@ -0,0 +1,265 @@
...
+      } else if (i == 1 || i == this.totalPages) {
+        this.listitems.push(this.configurePaginationStep(i));
+      } else if (i >= (this.currentPage-2) && i <= (this.currentPage+2)) {
+        this.listitems.push(this.configurePaginationStep(i));
+      } else if ((this.currentPage <= 5 && i <= 7) || (this.currentPage >= this.totalPages - 4 && i >= this.totalPages - 6)) {
+        this.listitems.push(this.configurePaginationStep(i));
+      } else if ((this.currentPage == 5 && i == 2) || (this.currentPage == this.totalPages -4 && i == this.totalPages - 1)) {
+        this.listitems.push(this.configurePaginationStep(i));
+      } else if (this.totalPages < 10) {
+        this.listitems.push(this.configurePaginationStep(i));
```
Comment (melaniebmn) [inline packages/web/src/components/gcds-pagination/gcds-pagination.tsx:166]: Looks like all these if statements up until here push the same configurePaginationStep, could you summarize them as one if statement or condense them a little?

**Review by melaniebmn — CHANGES_REQUESTED** [review]
Nice work removing a bunch of your JS code! Left a few comments, most of them are nit-picky formatting stuff :zany_face:

**Review by melaniebmn — APPROVED** [review]
LGTM
